README.md
postcss-color-hexa [![Build Status](https://travis-ci.org/nicksheffield/postcss-color-hexa.svg?branch=master)](https://travis-ci.org/nicksheffield/postcss-color-hexa)
===================
A postcss plugin that enables the use of hexa() to combine hex codes and opacity into rgba()

---

Input
```css
body {
	background: hexa(#abc, 0.1);
}
```

Output
```css
body {
	background: rgba(170, 187, 204, 0.1);
}
```

## Usage

```javascript
postcss([require('postcss-color-hexa')]);
```

CHANGELOG.md
## 1.2.3
* Allows for spaces everywhere! ie. `hexa(  #fff  ,  0.2  )`

## 1.2.2
* Allows for spaces after the opacity number

## 1.2.1
* Stopped infinite loops occuring
* Now throws errors on invalid hex colors and opacity numbers

## 1.2
* Initial release

test.js
var hexa = require('../hexa');
var postcss = require('postcss');
var expect = require('chai').expect;


describe('postcss-color-hexa', function() {

	function transform(source) {
		return postcss([hexa]).process(source).css;
	}

	it('short hex and full decimal', function(done) {
		expect(transform('body { background: hexa(#fff, 0.1); }'))
			.to.equal('body { background: rgba(255, 255, 255, 0.1); }');

		done();
	});

	it('full hex and full decimal', function(done) {
		expect(transform('body { background: hexa(#ffffff, 0.1); }'))
			.to.equal('body { background: rgba(255, 255, 255, 0.1); }');

		done();
	});

	it('short hex and integer', function(done) {
		expect(transform('body { background: hexa(#fff, 1); }'))
			.to.equal('body { background: rgba(255, 255, 255, 1); }');

		done();
	});

	it('full hex and integer', function(done) {
		expect(transform('body { background: hexa(#ffffff, 1); }'))
			.to.equal('body { background: rgba(255, 255, 255, 1); }');

		done();
	});

	it('short hex and zeroless decimal', function(done) {
		expect(transform('body { background: hexa(#fff, .1); }'))
			.to.equal('body { background: rgba(255, 255, 255, .1); }');

		done();
	});

	it('full hex and zeroless decimal', function(done) {
		expect(transform('body { background: hexa(#ffffff, .1); }'))
			.to.equal('body { background: rgba(255, 255, 255, .1); }');

		done();
	});

	it('hex with zeroes at the end and zeroless decimal', function(done) {
		expect(transform('body { background: hexa(#f00, .1); }'))
			.to.equal('body { background: rgba(255, 0, 0, .1); }');

		done();
	});

	it('hex with zeroes at the end and full decimal', function(done) {
		expect(transform('body { background: hexa(#f00, 0.1); }'))
			.to.equal('body { background: rgba(255, 0, 0, 0.1); }');

		done();
	});

	it('without hexa', function(done) {
		expect(transform('body { background: #f00; }'))
			.to.equal('body { background: #f00; }');

		done();
	});

	it('first within a multi-value property', function(done) {
		expect(transform('body { border: hexa(#fff, 0.1) 1px solid; }'))
			.to.equal('body { border: rgba(255, 255, 255, 0.1) 1px solid; }');

		done();
	});

	it('within a multi-value property', function(done) {
		expect(transform('body { border: 1px hexa(#fff, 0.1) solid; }'))
			.to.equal('body { border: 1px rgba(255, 255, 255, 0.1) solid; }');

		done();
	});

	it('last within a multi-value property', function(done) {
		expect(transform('body { border: 1px solid hexa(#fff, 0.1); }'))
			.to.equal('body { border: 1px solid rgba(255, 255, 255, 0.1); }');

		done();
	});

	it('multiple colors on a line including a hexa', function(done) {
		expect(transform('body { border-color: #fff hexa(#fff, 0.1) red #f00; }'))
			.to.equal('body { border-color: #fff rgba(255, 255, 255, 0.1) red #f00; }');

		done();
	});

	it('multiple hexa\'s on a line', function(done) {
		expect(transform('body { border-color: #0f0 hexa(#f0f, 0.1) hexa(#fff, 0.1) #f00; }'))
			.to.equal('body { border-color: #0f0 rgba(255, 0, 255, 0.1) rgba(255, 255, 255, 0.1) #f00; }');

		done();
	});

	it('invalid hex code fails', function(done) {
		// http://stackoverflow.com/questions/21587122/mocha-chai-expect-to-throw-not-catching-thrown-errors
		expect(function() {
			transform('body { background: hexa(#f0f0, 0.1); }');
		})
			.to.throw(/Invalid hex color/);

		done();
	});

	it('invalid opacity number fails', function(done) {
		expect(function() {
			transform('body { background: hexa(#f0f, 0.1.1); }');
		})
			.to.throw(/Invalid opacity/);

		done();
	});

});

test/test.js
var hexa = require('../hexa');
var postcss = require('postcss');
var chai = require('chai');

var expect = chai.expect;
var assert = chai.assert;


describe('postcss-color-hexa', function() {

	function transform (source) {
		var res;
		try {
			res = postcss([hexa]).process(source).css;
		} catch (err){
			// console.log(err);
			throw err;
			// res = err.reason;
		}
		return res;
	}

	it('short hex and full decimal', function(done) {
		expect(transform('body { background: hexa(#fff, 0.1); }'))
		.to.equal('body { background: rgba(255, 255, 255, 0.1); }');
		
		done();
	});
	
	it('full hex and full decimal', function(done) {
		expect(transform('body { background: hexa(#ffffff, 0.1); }'))
		.to.equal('body { background: rgba(255, 255, 255, 0.1); }');
		
		done();
	});

	it('short hex and integer', function(done) {
		expect(transform('body { background: hexa(#fff, 1); }'))
		.to.equal('body { background: rgba(255, 255, 255, 1); }');
		
		done();
	});
	
	it('full hex and integer', function(done) {
		expect(transform('body { background: hexa(#ffffff, 1); }'))
		.to.equal('body { background: rgba(255, 255, 255, 1); }');
		
		done();
	});
	
	it('short hex and zeroless decimal', function(done) {
		expect(transform('body { background: hexa(#fff, .1); }'))
		.to.equal('body { background: rgba(255, 255, 255, .1); }');
		
		done();
	});
	
	it('full hex and zeroless decimal', function(done) {
		expect(transform('body { background: hexa(#ffffff, .1); }'))
		.to.equal('body { background: rgba(255, 255, 255, .1); }');
		
		done();
	});
	
	it('hex with zeroes at the end and zeroless decimal', function(done) {
		expect(transform('body { background: hexa(#f00, .1); }'))
		.to.equal('body { background: rgba(255, 0, 0, .1); }');
		
		done();
	});

	it('hex with zeroes at the end and full decimal', function(done) {
		expect(transform('body { background: hexa(#f00, 0.1); }'))
		.to.equal('body { background: rgba(255, 0, 0, 0.1); }');
		
		done();
	});

	it('without hexa', function(done) {
		expect(transform('body { background: #f00; }'))
		.to.equal('body { background: #f00; }');
		
		done();
	});

	it('first within a multi-value property', function(done) {
		expect(transform('body { border: hexa(#fff, 0.1) 1px solid; }'))
		.to.equal('body { border: rgba(255, 255, 255, 0.1) 1px solid; }');
		
		done();
	});

	it('within a multi-value property', function(done) {
		expect(transform('body { border: 1px hexa(#fff, 0.1) solid; }'))
		.to.equal('body { border: 1px rgba(255, 255, 255, 0.1) solid; }');
		
		done();
	});

	it('last within a multi-value property', function(done) {
		expect(transform('body { border: 1px solid hexa(#fff, 0.1); }'))
		.to.equal('body { border: 1px solid rgba(255, 255, 255, 0.1); }');
		
		done();
	});

	it('multiple colors on a line including a hexa', function(done) {
		expect(transform('body { border-color: #fff hexa(#fff, 0.1) red #f00; }'))
		.to.equal('body { border-color: #fff rgba(255, 255, 255, 0.1) red #f00; }');
		
		done();
	});

	it('multiple hexa\'s on a line', function(done) {
		expect(transform('body { border-color: #0f0 hexa(#f0f, 0.1) hexa(#fff, 0.1) #f00; }'))
		.to.equal('body { border-color: #0f0 rgba(255, 0, 255, 0.1) rgba(255, 255, 255, 0.1) #f00; }');
		
		done();
	});

	it('invalid hex code fails', function(done) {
		// http://stackoverflow.com/questions/21587122/mocha-chai-expect-to-throw-not-catching-thrown-errors
		expect(function(){
			transform('body { background: hexa(#f0f0, 0.1); }');
		})
		.to.throw(/Invalid hex color/);
		
		done();
	});

	it('invalid opacity number fails', function(done) {
		expect(function(){
			transform('body { background: hexa(#f0f, 0.1.1); }');
		})
		.to.throw(/Invalid opacity/);
		
		done();
	});

});

test/index.js
var hexa = require('../hexa');
var postcss = require('postcss');


describe('hexa()', function(){
	
	it('short hex and full decimal', function(done){
		
		postcss([hexa()])
	
		.process('body { background: hexa(#fff, 0.1); }')
		
		.then(function (compiled) {
			if(compiled.css === 'body { background: rgba(255, 255, 255, 0.1); }'){
				done();
			}else{
				throw 'Output was incorrect';
			}
		});
	
	});
	
	it('full hex and full decimal', function(done){
		
		postcss([hexa()])
		
		.process('body { background: hexa(#ffffff, 0.1); }')
		
		.then(function (compiled) {
			if(compiled.css === 'body { background: rgba(255, 255, 255, 0.1); }'){
				done();
			}else{
				throw 'Output was incorrect';
			}
		});
		
	});
	
	it('short hex and integer', function(done){
		
		postcss([hexa()])
		
		.process('body { background: hexa(#fff, 1); }')
		
		.then(function (compiled) {
			if(compiled.css === 'body { background: rgba(255, 255, 255, 1); }'){
				done();
			}else{
				throw 'Output was incorrect';
			}
		});
		
	});
	
	it('full hex and integer', function(done){
		
		postcss([hexa()])
		
		.process('body { background: hexa(#ffffff, 1); }')
		
		.then(function (compiled) {
			if(compiled.css === 'body { background: rgba(255, 255, 255, 1); }'){
				done();
			}else{
				throw 'Output was incorrect';
			}
		});
		
	});
	
	
	
	it('short hex and zeroless decimal', function(done){
		
		postcss([hexa()])
		
		.process('body { background: hexa(#fff, .1); }')
		
		.then(function (compiled) {
			if(compiled.css === 'body { background: rgba(255, 255, 255, .1); }'){
				done();
			}else{
				throw 'Output was incorrect';
			}
		});
		
	});
	
	it('full hex and zeroless decimal', function(done){
		
		postcss([hexa()])
		
		.process('body { background: hexa(#ffffff, .1); }')
		
		.then(function (compiled) {
			if(compiled.css === 'body { background: rgba(255, 255, 255, .1); }'){
				done();
			}else{
				throw 'Output was incorrect';
			}
		});
		
	});
	
	it('hex with zeroes at the end and zeroless decimal', function(done){
		
		postcss([hexa()])
		
		.process('body { background: hexa(#f00, .1); }')
		
		.then(function (compiled) {
			if(compiled.css === 'body { background: rgba(255, 0, 0, .1); }'){
				done();
			}else{
				throw 'Output was incorrect';
			}
		});
		
	});
	
	it('hex with zeroes at the end and full decimal', function(done){
		
		postcss([hexa()])
		
		.process('body { background: hexa(#f00, 0.1); }')
		
		.then(function (compiled) {
			if(compiled.css === 'body { background: rgba(255, 0, 0, 0.1); }'){
				done();
			}else{
				throw 'Output was incorrect';
			}
		});
		
	});
	
	
	
	it('without hexa', function(done){
		
		postcss([hexa()])
		
		.process('body { background: #f00; }')
		
		.then(function (compiled) {
			if(compiled.css === 'body { background: #f00; }'){
				done();
			}else{
				throw 'Output was incorrect';
			}
		});
		
	});
	
	it('first within a multi-value property', function(done){
		
		postcss([hexa()])
		
		.process('body { border: hexa(#fff, 0.1) 1px solid; }')
		
		.then(function (compiled) {
			if(compiled.css === 'body { border: rgba(255, 255, 255, 0.1) 1px solid; }'){
				done();
			}else{
				throw 'Output was incorrect';
			}
		});
		
	});
	
	it('within a multi-value property', function(done){
		
		postcss([hexa()])
		
		.process('body { border: 1px hexa(#fff, 0.1) solid; }')
		
		.then(function (compiled) {
			if(compiled.css === 'body { border: 1px rgba(255, 255, 255, 0.1) solid; }'){
				done();
			}else{
				throw 'Output was incorrect';
			}
		});
		
	});
	
	it('last within a multi-value property', function(done){
		
		postcss([hexa()])
		
		.process('body { border: 1px solid hexa(#fff, 0.1); }')
		
		.then(function (compiled) {
			if(compiled.css === 'body { border: 1px solid rgba(255, 255, 255, 0.1); }'){
				done();
			}else{
				throw 'Output was incorrect';
			}
		});
		
	});
	
});

index.js
// ----------------------------------------------------------------------------
// Dependencies
// ----------------------------------------------------------------------------

var helpers = require("postcss-message-helpers");


// ----------------------------------------------------------------------------
// Regex's
// ----------------------------------------------------------------------------

// Match the whole property. I don't think I need this
var HEX_ALPHA_RE = /hexa\(#([0-9a-fA-F]{3,6}),([\s]{1,})([0-9\.]{1,})\)/i;

// Match the hex color
var HEX_RE = /#(?:[0-9a-f]{3}){1,2}/;

// Match the opacity number
var OPACITY_RE = /[0-9\.]{1,}(?=\))/;


// ----------------------------------------------------------------------------
// PostCSS Plugin
// ----------------------------------------------------------------------------

module.exports = function plugin() {
	return function(style) {
		style.eachDecl(function transformDecl(decl) {
			if (!decl.value || decl.value.indexOf("hexa") === -1) {
				return;
			}

			decl.value = helpers.try(function transformHexAlphaValue() {
				return transformHexAlpha(decl.value, decl.source);
			}, decl.source);
		});
	};
};


// ----------------------------------------------------------------------------
// Helper functions
// ----------------------------------------------------------------------------

function transformHexAlpha(string) {
	var h = HEX_RE.exec(string);
	var o = OPACITY_RE.exec(string);
	if (!h) {
		return string;
	}
	
	var rgb = hexToRgb(h[0]);
	var rgb_chars = [rgb.r, rgb.g, rgb.b];
	
	if(o){
		rgb_chars.push(o);
		return 'rgba(' + rgb_chars.join(', ') +')';
	} else {
		return 'rgb(' + rgb_chars.join(', ') + ')';
	}
}

// http://stackoverflow.com/questions/5623838/rgb-to-hex-and-hex-to-rgb
function hexToRgb(hex) {
    // Expand shorthand form (e.g. "03F") to full form (e.g. "0033FF")
    var shorthandRegex = /^#?([a-f\d])([a-f\d])([a-f\d])$/i;
    hex = hex.replace(shorthandRegex, function(m, r, g, b) {
        return r + r + g + g + b + b;
    });

    var result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
    return result ? {
        r: parseInt(result[1], 16),
        g: parseInt(result[2], 16),
        b: parseInt(result[3], 16)
    } : null;
}

hexa.js
// ----------------------------------------------------------------------------
// Dependencies
// ----------------------------------------------------------------------------

var helpers = require("postcss-message-helpers");
var postcss = require("postcss");


// ----------------------------------------------------------------------------
// Regex's
// ----------------------------------------------------------------------------

// Match the whole property. Test: http://regexr.com/3bep1
var HEX_ALPHA_RE = /hexa\(\s*#(?:[0-9a-f]{3}){1,2}\s*,[\s]{0,}[0-9]*\.?[0-9]+\s*\)/i;

// Match the hex color. Test: http://regexr.com/3beoo
var HEX_RE = /#(?:[0-9a-f]{3}){1,2}/i;

// Match the opacity number. Test: http://regexr.com/3beol
var OPACITY_RE = /[0-9]*(\.)?[0-9]+(?=[\s]*\))/;

// Match an invalid hex code (followed by a comma). Test: http://regexr.com/3beoi
var INV_HEX_RE = /#(?:[0-9a-f]{0,2}|[0-9a-f]{4,5}|[0-9a-f]{7,})(?=,)/i;

// Match an invalid opacity number. Test: http://regexr.com/3beof
var INV_OPACITY_RE = /[0-9]*\.?[0-9]*(?:\.[0-9]*){2,}/i;

// ----------------------------------------------------------------------------
// PostCSS Plugin
// ----------------------------------------------------------------------------

module.exports = postcss.plugin('postcss-color-hexa', function(opts) {
	opts = opts || {};

	return function(css) {
		css.eachDecl(function transformDecl(decl) {
			if (!decl.value || decl.value.indexOf('hexa') === -1) {
				return;
			}

			decl.value = helpers.try(function transformHexAlphaValue() {
				return transformHexAlpha(decl.value, decl.source, decl);
			}, decl.source);
		});
	};
});


// ----------------------------------------------------------------------------
// Helper functions
// ----------------------------------------------------------------------------

function transformHexAlpha(string, source, decl) {

	return recurse(string);

	function recurse(input) {
		var hexa_pos = input.indexOf('hexa');
		var subs = input.substring(hexa_pos);
		
		var h = HEX_RE.exec(subs);
		var o = OPACITY_RE.exec(subs);
		
		var i_h = INV_HEX_RE.exec(subs);
		var i_o = INV_OPACITY_RE.exec(subs);
		
		if(i_h !== null){
			throw decl.error('Invalid hex color: ' + i_h[0], { plugin: 'postcss-color-hexa' });
		}
		
		if(i_o !== null){
			throw decl.error('Invalid opacity: ' + i_o[0], { plugin: 'postcss-color-hexa' });
		}

		if (!h) {
			return input;
		}

		var rgb = hexToRgb(h[0]);
		var rgb_chars = [rgb.r, rgb.g, rgb.b];
		var hexa = '';

		if (o) {
			rgb_chars.push(o[0]);
			hexa = 'rgba(' + rgb_chars.join(', ') + ')';
		} else {
			hexa = 'rgb(' + rgb_chars.join(', ') + ')';
		}

		var result = input.replace(HEX_ALPHA_RE, hexa);

		if (result.indexOf('hexa') === -1) {
			return result;
		} else {
			return recurse(result);
		}
	}
}

// http://stackoverflow.com/questions/5623838/rgb-to-hex-and-hex-to-rgb
function hexToRgb(hex) {
	// Expand shorthand form (e.g. "03F") to full form (e.g. "0033FF")
	var shorthandRegex = /^#?([a-f\d])([a-f\d])([a-f\d])$/i;
	hex = hex.replace(shorthandRegex, function(m, r, g, b) {
		return r + r + g + g + b + b;
	});

	var result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
	return result ? {
		r: parseInt(result[1], 16),
		g: parseInt(result[2], 16),
		b: parseInt(result[3], 16)
	} : null;
}